src/tree/split_evaluator.cc

```diff
@@ -303,5 +306,195 @@ XGBOOST_REGISTER_SPLIT_EVALUATOR(MonotonicConstraint, "monotonic")
+                bst_uint rightid,
+                bst_uint featureid,
+                bst_float leftweight,
+                bst_float rightweight) override {
+    inner_->AddSplit(nodeid, leftid, rightid, featureid, leftweight, rightweight);
+
+    if (params_.interaction_constraints.empty()) {
+      return;  // short-circuit if no constraint is specified
+    }
+    bst_uint newsize = std::max(leftid, rightid) + 1;
+
+    // Record previous splits for child nodes
+    std::unordered_set<bst_uint> feature_splits = splits_[nodeid];  // fid history of current node
+    feature_splits.insert(featureid);  // add feature of current node
+    splits_.resize(newsize);
+    splits_[leftid] = feature_splits;
+    splits_[rightid] = feature_splits;
+
+    // Resize constraints record, initialise all features to be not permitted for new nodes
+    int_cont_.resize(newsize, std::unordered_set<bst_uint>());
+
+    // Permit features used in previous splits
+    for (bst_uint fid : feature_splits) {
+      int_cont_[leftid].insert(fid);
+      int_cont_[rightid].insert(fid);
+    }
+
+    // Loop across specified interactions in constraints
+    for (const auto& constraint : interaction_constraints_) {
+      bst_uint flag = 1;  // flags whether the specified interaction is still relevant
+
+      // Test relevance of specified interaction by checking all previous features are included
+      for (bst_uint checkvar : feature_splits) {
+        if (constraint.count(checkvar) == 0) {
+          flag = 0;
+          break;  // interaction is not relevant due to unmet constraint
+        }
+      }
+
+      // If interaction is still relevant, permit all other features in the interaction
+      if (flag == 1) {
+        for (bst_uint k : constraint) {
+          int_cont_[leftid].insert(k);
+          int_cont_[rightid].insert(k);
+        }
+      }
+    }
+  }
+
+ private:
+  InteractionConstraintParams params_;
+  std::unique_ptr<SplitEvaluator> inner_;
+  // interaction_constraints_[constraint_id] contains a single interaction
+  //   constraint, which specifies a group of feature IDs that can interact
+  //   with each other
+  std::vector< std::unordered_set<bst_uint> > interaction_constraints_;
+  // int_cont_[nid] contains the set of all feature IDs that are allowed to
+  //   be used for a split at node nid
+  std::vector< std::unordered_set<bst_uint> > int_cont_;
+  // splits_[nid] contains the set of all feature IDs that have been used for
+  //   splits in node nid and its parents
+  std::vector< std::unordered_set<bst_uint> > splits_;
+
+  // Check interaction constraints. Returns true if a given feature ID is
+  //   permissible in a given node; returns false otherwise
+  inline bool CheckInteractionConstraint(bst_uint featureid, bst_uint nodeid) const {
+    if (params_.interaction_constraints.empty()) {
+      return true;  // short-circuit if no constraint is specified
+    }
+    return (int_cont_[nodeid].count(featureid) > 0);
```

Comment:
return (params_.interaction_constraints.empty() || int_cont_[nodeid].count(featureid) > 0);
would have more chance of inlining

Reply:
Fixed